vita3k/util/src/net_utils.cpp

```diff
@@ -393,47 +393,58 @@ bool download_file(const std::string &url, const std::string &output_file_path,
     if (!curl_download)
         return false;
 
-    CurlDownloadCallback curl_callback = +[](void *user_data, long bytes_total, long downloaded_bytes) {
+    CurlDownloadCallback curl_callback = +[](void *user_data, long total_bytes, long downloaded_bytes) {
         const auto data = (CallbackData *)user_data;
 
-        if (bytes_total == 0)
-            return 0; // Ignore if we dont have the total yet
+        if ((total_bytes == 0) || (downloaded_bytes == 0))
+            return 0; // Ignore if we dont have the total yet or if we haven't downloaded anything
 
         if (!data->second)
             return 0;
-        const auto progress_percent = static_cast<float>(downloaded_bytes) / static_cast<float>(bytes_total) * 100.0f;
 
-        const auto elapsed_time_ms = std::difftime(get_current_time_ms(), data->first);
+        // Calculate progress percentage
+        const auto total_bytes_downloaded = static_cast<double>(downloaded_bytes + data->first.bytes_already_downloaded);
+        const auto file_size = total_bytes + data->first.bytes_already_downloaded;
+        const auto progress_percent = static_cast<float>(total_bytes_downloaded) / static_cast<float>(file_size) * 100.0f;
+
+        // Calculate elapsed time
+        const auto current_time = get_current_time_ms();
+        const auto elapsed_time_ms = std::difftime(current_time, data->first.time);
 
         // Calculate remaining time in seconds
-        const auto remaining_bytes = static_cast<double>(bytes_total - downloaded_bytes);
+        const auto remaining_bytes = static_cast<double>(total_bytes - downloaded_bytes);
         const auto remaining_time = static_cast<uint64_t>((remaining_bytes / downloaded_bytes) * elapsed_time_ms) / 1000;
 
         ProgressState *callback_result = data->second(progress_percent, remaining_time);
 
         std::unique_lock<std::mutex> lock(callback_result->mutex);
 
-        callback_result->cv.wait(lock, [&]() {
-            return !callback_result->pause;
-        });
+        if (callback_result->pause) {
+            callback_result->cv.wait(lock);
```

Comment:
Don't wait without a predicate, c++ condition variables are allowed to have a spurious wakeup

Reply:
yes, i have fixed it withlittle reworks this part